packages/storybook-react/stories/extension-entity-reference/highlights.tsx

```diff
@@ -0,0 +1,106 @@
+import 'remirror/styles/all.css';
+import './styles.css';
+
+import { cx, htmlToProsemirrorNode, uniqueId } from 'remirror';
+import {
+  EntityReferenceExtension,
+  EntityReferenceMetaData,
+  findMinMaxRange,
+} from 'remirror/extensions';
+import { Decoration } from '@remirror/pm/view';
+import { Remirror, ThemeProvider, useCommands, useHelpers, useRemirror } from '@remirror/react';
+
+type HighlightTypes = 'important' | 'interesting';
+const ALL_HIGHLIGHT_TYPES: HighlightTypes[] = ['important', 'interesting'];
+
+const allHighlights = new Map<string, HighlightTypes>();
+
+export const decorateHighlights = (highlights: EntityReferenceMetaData[][]): Decoration[] => {
+  const decorations = highlights.map((overlappingHighlights) => {
+    const types = new Set(overlappingHighlights.map((h) => allHighlights.get(h.id)));
+    // Mix colors to allow for overlapping highlights
+    const notRed = types.has('important') ? 64 : 0;
+    const notBlue = types.has('interesting') ? 64 : 0;
+    const red = 255 - notBlue;
+    const green = 255 - notBlue - notRed;
+    const blue = 255 - notRed;
+    const style = `background: rgb(${red}, ${green}, ${blue});padding: 6px 0;`;
+    const [from, to] = findMinMaxRange(overlappingHighlights);
+
+    // Add decoration to all inline nodes in the given range.
+    return Decoration.inline(from, to, { style });
+  });
+
+  return [...decorations];
+};
+
+const extensions = () => [
+  new EntityReferenceExtension({
+    getStyle: decorateHighlights,
+  }),
+];
+
+const Buttons = () => {
+  const { getEntityReferencesAt } = useHelpers<EntityReferenceExtension>(true);
+  const commands = useCommands<EntityReferenceExtension>();
+  const highlightsAt = getEntityReferencesAt();
+  return (
+    <>
+      {ALL_HIGHLIGHT_TYPES.map((type) => {
+        const highlightsOfType = highlightsAt.filter((h) => {
+          const highlightType = allHighlights.get(h.id);
+          return highlightType === type;
+        });
+        // Provide visual feedback if there is a highlight of this type at the user's cursor
+        const active = highlightsOfType.length > 0;
```

Comment:
Just curious about this check, it means you can't nested entity references. Is feel like it's misleading.

Assuming text formatting here as entity references (sorry best I can do in markdown)

_A paragraph of text with a **nested entity reference** within it_

Reply:
You can. For example the storybook (cursor is on "and"):
![image](https://user-images.githubusercontent.com/9339055/173194413-7679a2e1-13b0-4ead-846d-d29ae3a8f791.png)

Reply:
But you can't use the "important" tag twice. The second time removes the first one.

Reply:
Aha, now I get it. Indeed, nesting isn't possible in the example. I didn't really know how to keep the example minimal otherwise (which would require multiple separate remove buttons).
You think the demo still brings the point across? Shall we just add a comment to point out this behavior? Or you think the storybook should be different (e.g. have remove buttons instead of just toggle button)?

Reply:
A remove button could work. I was also wondering if we could compare the start and end positions of an existing reference too?

I.e. if the positions are same, delete it

Reply:
I thought about this a bit further, and I increasingly think that we shouldn't change the storybook because:

**1. For highlights (ie. the storybook example)**
We know from our own usage over the last years that this a theoretical option. We haven't seen any user highlight a sentence as "important" and then as well a word within the same sentence as "important". Users might highlight nested content with different entities/tags - but with the same is quite a corner case. 
Making removing a separate actions would make the storybook considerable more complex because we have to handle than also "which highlight of this type should be removed", etc.
The main purpose of the storybooks is to demonstrate common usage in the simplest possible way (e.g. without fancy styling, error handling, etc). My feeling is that we would serve this here better by not covering the corner case.

**2. For non-highlights:**
I'm not sure if referencing the same entity from nested content is relevant or not. We haven't see it but I would also not rule it out. So, I'd keep it as a possibility in the extension (aside that forbidding would actually make the extension more complex) but wait with adding a storybook for that case until we have a real-world use case.

Does this reasoning make sense to you?

> I was also wondering if we could compare the start and end positions of an existing reference too?

Comparing start/end position is tricky because references might have "holes". See below an example from Google Docs where the yellow comment is split in two parts:
![image](https://user-images.githubusercontent.com/9339055/173222823-3f415a1f-bcc7-4a41-8005-232893a6a0c6.png)

These holes can happen if you copy & paste parts of the text with a mark. One could do funky things like also copying the entities (e.g. one comment become two comments) but this isn't always what the user expects ("Why do I have now two comments?") and it adds a lot of follow up complexity (e.g. undo/redo required to split/re-join entities).

Reply:
Totally get your point on number 1 and agree we should keep it simple.

For non highlight entities, (specifically comments) I imagined a paragraph of text with bad grammar, and a single misspelled word.

I imagine a user selecting the entire paragraph and commenting. 
Then, selecting the misspelled word (nested in the paragraph) and adding another comment on that. 

In this use case they would have different entity IDs so maybe, this is already covered?

Reply:
Yes, indeed. Different comments would be different entities.